src/v/kafka/server/handlers/fetch.cc

```diff
@@ -671,6 +672,477 @@ class parallel_fetch_plan_executor final : public fetch_plan_executor::impl {
+                fetch.responses.erase(fetch.responses.begin() + (i - shift));
+                shift += 1;
+            }
+        }
+
+        _last_result_size[shard] = results.total_size;
+        _completed_shard_fetches.push_back(std::move(fetch));
+        _has_data.signal();
+    }
+
+    std::optional<ss::future<>> register_hwm_wait(
+      shard_local_fetch_context& ctx,
+      const model::ktp& ktp,
+      model::offset offset,
+      ss::abort_source& as) {
+        auto part = ctx.mgr.get(ktp);
+        // If the partition can't be found then it's since been moved
+        if (!part) {
+            return {};
+        }
+
+        auto consensus = part->raft();
+        if (!consensus) {
+            return {};
+        }
+
+        return {ignore_exceptions(consensus->visible_offset_monitor().wait(
+          offset, model::no_timeout, as))};
+    }
+
+    /**
+     * The shard fetch worker will first attempt to query every partition in
+     * `fetch` for data.
+     *
+     * If there is enough data to meet `min_fetch_bytes` or if
+     * the fetch coordinator has sent an abort then the worker returns.
+     *
+     * Otherwise it'll wait for the last_visible_index to change on one of the
+     * partitions in `fetch` then attempt to gather enough data again.
+     *
+     * This will repeat until `min_fetch_bytes` is met or an abort is received
+     * from the coordinator.
+     */
+    ss::future<worker_result>
+    shard_fetch_worker(shard_local_fetch_context ctx) {
+        size_t total_size = 0;
+        std::vector<size_t> errored_partitions;
+
+        // The last visible indexes need to be populated before partitions are
+        // read. If they are populated afterwards then the last_visible_index
+        // could be updated after the partition is read, but before the index is
+        // queried. This leads to a race condition where we're waiting for the
+        // last_visible_index to change even though the partition has data we
+        // haven't read.
+        //
+        // Note that `last_visible_index` is used instead of the `commit_offset`
+        // as the commit offset won't change when partitions are produced to
+        // without acks=all. The `last_visible_index` more closely corresponds
+        // to the Kafka high watermark as well.
+        std::vector<model::offset> last_visible_indexes{ctx.requests.size()};
+
+        for (size_t i = 0; i < ctx.requests.size(); i++) {
+            const auto& req = ctx.requests[i];
+            auto part = ctx.mgr.get(req.ktp());
+            if (!part) {
+                errored_partitions.push_back(i);
+                continue;
+            }
+            auto consensus = part->raft();
+            if (!consensus) {
+                errored_partitions.push_back(i);
+                continue;
+            }
+            last_visible_indexes[i] = consensus->last_visible_index();
+        }
+
+        // Fast-path fetch
```

Comment:
question, why is this necessary? I would normally just expect a loop executing this, then resetting the promise when exhausted, why does this need to exist out of that loop?

Reply:
It originally avoided a few allocations. However, it wasn't really optimal and the most recent version of this PR doesn't have it.